sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -987,10 +987,10 @@ object CollapseRepartition extends Rule[LogicalPlan] {
       case (false, true) => if (r.numPartitions >= child.numPartitions) child else r
       case _ => r.copy(child = child.child)
     }
-    // Case 2: When a RepartitionByExpression has a child of Repartition or RepartitionByExpression
-    // we can remove the child.
-    case r @ RepartitionByExpression(_, child: RepartitionOperation, _) =>
-      r.copy(child = child.child)
+    // Case 2: When a RepartitionByExpression has a child of Sort, Repartition or
+    // RepartitionByExpression we can remove the child.
+    case r @ RepartitionByExpression(_, child @ (_: Sort | _: RepartitionOperation), _) =>
```

Comment:
is it possible that Spark can keep the ordering in some cases?

Reply:
e.g. partition key is the same as sort key?

Reply:
Do you mean `spark.range(10).selectExpr("cast(id AS string)").repartition($"id").sort("id")`?

Reply:
e.g. `spark.table(...).repartition("a").sortWithinPartitions("a").repartition(a)`.

Reply:
ok. make it only remove the global sort.